vault/namespaces_store.go

```diff
@@ -755,3 +773,176 @@ func (ns *NamespaceStore) ResolveNamespaceFromRequest(nsHeader, reqPath string)
 
 	return resolvedNs, trimmedPath
 }
+
+// GetLockingNamespace walks the namespace tree structure looking for
+// a locked namespace, starting from one of the root children,
+// ending at the namespace provided as a argument to the function.
+func (ns *NamespaceStore) GetLockingNamespace(n *namespace.Namespace) *namespace.Namespace {
+	var lockedNS *namespace.Namespace
+	ns.namespacesByPath.WalkPath(n.Path, func(curNS *namespace.Namespace) bool {
+		if curNS.Locked {
+			lockedNS = curNS
+			return true
+		}
+		return false
+	})
+	if lockedNS != nil {
+		return lockedNS.Clone()
+	}
+
+	return nil
+}
+
+// UnlockNamespace attempts to unlock the namespace with provided provided uuid.
+func (ns *NamespaceStore) UnlockNamespace(ctx context.Context, unlockKey, path string, forceUnlock bool) error {
+	defer metrics.MeasureSince([]string{"namespace", "unlock_namespace"}, time.Now())
+
+	if err := ns.checkInvalidation(ctx); err != nil {
+		return err
+	}
+
+	namespaceToUnlock, err := ns.GetNamespaceByPath(ctx, path)
+	if err != nil {
+		return err
+	}
+
+	if namespaceToUnlock == nil {
+		return errors.New("requested namespace does not exist")
+	}
+
+	if !namespaceToUnlock.Locked {
+		return fmt.Errorf("namespace %q is not locked", namespaceToUnlock.Path)
+	}
```

Comment:
```suggestion

	if namespaceToUnlock.ID == namespace.RootNamespaceID {
		return fmt.Errorf("cannot unlock root namespace")
	}

	if !namespaceToUnlock.Locked {
		return fmt.Errorf("namespace %q is not locked", namespaceToUnlock.Path)
	}
```

else we get `* namespace "" is not locked` for a simple `bao namespace unlock`.

Reply:
good catch

Reply:
@satoqz 
I wonder if we should respond with error like that here, it's not possible to lock `root` namespace, so trying to unlock it is completely invalid operation, and the response of `cannot unlock root namespace` could imply that it is locked and cannot be unlocked for some reason...